pkg/netutil/netutil.go

```diff
@@ -174,29 +174,29 @@ func URLStringsEqual(ctx context.Context, lg *zap.Logger, a []string, b []string
 	if len(a) != len(b) {
 		return false, fmt.Errorf("len(%q) != len(%q)", a, b)
 	}
-	urlsA := make([]url.URL, 0)
-	for _, str := range a {
-		u, err := url.Parse(str)
-		if err != nil {
-			return false, fmt.Errorf("failed to parse %q", str)
-		}
-		urlsA = append(urlsA, *u)
+	urlsA, err := stringsToURLs(a)
+	if err != nil {
+		return false, err
 	}
-	urlsB := make([]url.URL, 0)
-	for _, str := range b {
-		u, err := url.Parse(str)
-		if err != nil {
-			return false, fmt.Errorf("failed to parse %q", str)
-		}
-		urlsB = append(urlsB, *u)
+	urlsB, err := stringsToURLs(b)
+	if err != nil {
+		return false, err
 	}
 	if lg == nil {
 		lg, _ = zap.NewProduction()
 		if lg == nil {
 			lg = zap.NewExample()
 		}
 	}
-	return urlsEqual(ctx, lg, urlsA, urlsB)
+	sort.Sort(types.URLs(urlsA))
+	sort.Sort(types.URLs(urlsB))
+	for i := range urlsA {
+		if !reflect.DeepEqual(urlsA[i], urlsB[i]) {
```

Comment:
The idea behind this PR is just not to go down to the resolver if urls are equal as strings.
The bootstrap code is not performance critical, so here the use of `reflect.DeepEqual` is similar to `a.String () == b.String ()`, this piece is copied from [netutil/netutil.go#L160](https://github.com/etcd-io/etcd/blob/b922a286220f2348ca8662dbf8dc7203cde57f66/pkg/netutil/netutil.go#L160)

If the code immediately goes down to resolving, then bootstrap slows down by the timeout of the resolving context, which significantly slows down the start of the node.

If two urls are equal as strings, there is no point in resolving them, because they must be resolved to the same ip addresses, an exception may be the dns name somehow tied to a certain set of ip and resolving only to one of these ip, but in this case it is better to avoid resolving.

From tests and context of use, I came to the conclusion that the main purpose of the `URLStringsEqual` function is to return true if there is a comparison between `http://node1.com` and `http://1.2.3.4` where `node1.com=1.2.3.4`

Reply:
Turns out this comment from Lili was spot on, with the Hypershift product that recently upgraded to the latest 3.5 release including this fix, we've seen troubles with wildcard dns names matching the server name early:

> "*.etcd-discovery.e2e-clusters-8f75l-example-54cx2.svc"
> etcd-0.etcd-discovery.e2e-clusters-8f75l-example-54cx2.svc

which in turn caused issues on connecting through the TLS listener. Previously the resolver would block the bootstrap until the DNS resolves. 

The team has created an init container to wait for the resolution before starting etcd: https://github.com/openshift/hypershift/pull/1985

Reply:
thx @tjungblu , https://github.com/etcd-io/etcd/issues/15062